modules/reactor/src/meshgenerators/HexIDPatternedMeshGenerator.C

```diff
@@ -0,0 +1,149 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "HexIDPatternedMeshGenerator.h"
+
+registerMooseObject("ReactorApp", HexIDPatternedMeshGenerator);
+
+InputParameters
+HexIDPatternedMeshGenerator::validParams()
+{
+  InputParameters params = PatternedHexMeshGenerator::validParams();
+  params.addRequiredParam<std::string>("id_name", "Reporting_id_name");
+  params.addParam<std::vector<MeshGeneratorName>>(
+      "exclude_id", "Name of inputs to be excluded in ID generation.");
+  MooseEnum option("cell pattern manual", "cell");
+  params.addParam<MooseEnum>("assign_type", option, "Type of integer id assignment");
+  params.addParam<std::vector<std::vector<dof_id_type>>>(
+      "id_pattern",
+      "User-defined element IDs. A double-indexed array starting with the upper-left corner");
+  params.addClassDescription("This HexIDPatternedMeshGenerator source code is to generate "
+                             "patterned hexagonal meshes with Reporting ID");
+  return params;
+}
+
+HexIDPatternedMeshGenerator::HexIDPatternedMeshGenerator(const InputParameters & parameters)
+  : PatternedHexMeshGenerator(parameters),
+    _element_id_name(getParam<std::string>("id_name")),
+    _assign_type(getParam<MooseEnum>("assign_type")),
+    _use_exclude_id(isParamValid("exclude_id"))
+{
+  if (_use_exclude_id && _assign_type != "cell")
+    paramError("exclude_id", "works only when \"assign_type\" is equal 'cell'");
+  if (!isParamValid("id_pattern") && _assign_type == "manual")
+    paramError("id_pattern", "required when \"assign_type\" is equal to 'manual'");
+
+  if (_assign_type == "manual")
+    _id_pattern = getParam<std::vector<std::vector<dof_id_type>>>("id_pattern");
+  _exclude_ids.resize(_input_names.size());
+  if (_use_exclude_id)
+  {
+    std::vector<MeshGeneratorName> exclude_id_name =
```

Comment:
More duplicated code

Reply:
@loganharbour 
Duplications in Cartesian and hexagonal ones were removed by moving the those duplicated code sections into the utils. Thank you for suggestions.